server/spotify-connector.js

```diff
@@ -1,40 +1,44 @@
-require('dotenv').config()
-
 const request = require('request') // might change to axios
-const Spotify = require('spotify-web-api-js')
-
-const spotifyApi = new Spotify()
 
-const {
-  CLIENT_ID,
-  CLIENT_SECRET
-} = process.env
 const CLIENT_CREDENTIALS = 'client_credentials'
 
-const authOptions = {
-  url: 'https://accounts.spotify.com/api/token',
-  headers: {
-    'Authorization': `Basic ${Buffer.from(`${CLIENT_ID}:${CLIENT_SECRET}`).toString('base64')}`
-  },
-  form: {
-    grant_type: CLIENT_CREDENTIALS
-  },
-  json: true
-}
+class SpotifyConnector {
+
+  constructor(Config, SpotifyApi) {
+    this.Config = Config
+    this.SpotifyApi = SpotifyApi
 
-function obtainAccessToken(callback) {
-  request.post(authOptions, (error, res, body) => {
-    if (!error && res.statusCode === 200) {
-      const token = body.access_token
-      callback(null, token)
+    this.authOptions = {
+      url: 'https://accounts.spotify.com/api/token',
+      headers: {
+        'Authorization': `Basic ${Buffer.from(`${this.Config.spotifyAuth.clientId}:${this.Config.spotifyAuth.clientSecret}`).toString('base64')}`
+      },
+      form: {
+        grant_type: CLIENT_CREDENTIALS
+      },
+      json: true
     }
-  })
-}
+    this.refreshAccessToken()
+  }
+
+  getApi() {
+    return this.SpotifyApi
+  }
+
+  obtainAccessToken(callback) {
+    request.post(this.authOptions, (error, res, body) => {
+      if (!error && res.statusCode === 200) {
+        this.token = body.access_token
+        callback(null, this.token)
+      }
+    })
+  }
 
-function refreshAccessToken() {
-  obtainAccessToken((err, res) => {
-    spotifyApi.setAccessToken(res)
-  })
+  refreshAccessToken() {
+    this.obtainAccessToken((err, res) => {
+      this.SpotifyApi.setAccessToken(res)
```

Comment:
Hmm metody `refreshAccessToken` i `obtainAccessToken` są trochę mylące, bo o ile się nie mylę (;P) to samo `obtainAccessToken` nic nie robi? Nie do końca rozumiem rozbicia tego na 2 metody. I w konstruktorze od razu używasz `refreshAccessToken` co nie jest prawdą, bo nie refreshujesz tylko pobierasz pierwszy raz :P (chyba)

Reply:
`obtainAccessToken` wykonuje posta uzywajac mojego client_id i client_secret, w odpowiedzi na tego posta Spotify zwraca mi Access Token, ktory jest wazny na godzine. Po tej godzinie musze znowu zapytac Spotify o nowy Access Token, bo ten wygasa i ponownie ustawic go w SpotifyApi, zeby z niego korzystalo. Moglbym nazwac to `setAccessToken` - ale wlasnie `refreshAccessToken` wydawalo mi sie bardziej zrozumiale. Poniewaz `setAccessToken` wykonuje z wrappera, a caly proces refreshowania zawiera w sobie wykonanie posta i powiazanie tokena z api.

Reply:
No to rozumiem, ale po co masz `refreshAccessToken` zamiast po prostu w metodzie `obtainAccessToken` nie robić `callback(null, this.token)` tylko `this.SpotifyApi.setAccessToken(body.acess_token)`?

Reply:
No w sumie moge. Myslalem, ze w ten sposob jest bardziej SOLIDowe, bo mamy single responsibility. Ta metoda pobiera. Ta ustawia. Ale w sumie jest na tyle mala, ze chyba nie ma to znaczenia, wiec zmienilem. Zmienilem tez nazwe na setNewAccessToken - chyba pokrywa wszystkie case'y :D